Add dedicated type for storing image hashes

GnomedDev opened the pull request:
This converts the hex representation that is used across the API into a much smaller (56 vs 17) type which improves type safety, saves space, and reduces heap allocations (only during a clone, but deserialization could be made work without an allocation if we pull in an ArrayString dependency).

GnomedDev committed:
Add dedicated type for storing image hashes

GnomedDev committed:
Fix tests

GnomedDev committed:
Remove debug print

arqunis commented on src/model/guild/mod.rs:
```diff
@@ -626,7 +626,7 @@ impl Guild {
         http: impl AsRef<Http>,
         name: &str,
-        icon: Option<&str>,
+        icon: ImageHash,
```
Why was this changed to a plain `ImageHash`? Guilds can be created without an icon, so this should stay as `Option`al.

```suggestion
        icon: Option<ImageHash>,
```

GnomedDev replied:
Ah, guess this was a mistake

arqunis commented on src/model/misc.rs:
```diff
@@ -53,6 +51,108 @@ impl_from_str! {
+/// An image hash returned from the Discord API.
+///
+/// Note: This is parsed into a compact form when constructed, then turned back
+/// into the cannonical hex representation used by the API when needed.
```
Speaking of my previous comment, it is not immediately clear how this type should be constructed by the user if it is going to be passed down as an argument to functions like `Guild::create`. We should mention the `FromStr` implementation here with an example.

arqunis commented on src/model/misc.rs:
```diff
@@ -53,6 +51,108 @@ impl_from_str! {
+}
+
+#[derive(Debug, Clone)]
+pub enum ImageHashParseError {
```
If `FromStr` will be the go-to method for constructing an `ImageHash`, the error type should be documented as well.

arqunis commented on src/model/misc.rs:
```diff
@@ -53,6 +51,108 @@ impl_from_str! {
+                write!(f, "Invalid length {length}, expected 32 or 34 characters")
+            },
+            Self::MissingAnimatedMark => f.write_str("Input is 34 characters long, but missing a_"),
+            Self::UnparsableBytes(bytes) => write!(f, "Could not parse to hex: {bytes}"),
```
Types that implement `std::error::Error` should not be embedded within `Display`'d messages. Instead, they should be returned as [the `source` of the error](https://doc.rust-lang.org/nightly/std/error/trait.Error.html#method.source), so that error crates [like `color-eyre`](https://github.com/yaahc/color-eyre) can integrate with these types nicely. While we're on the topic, `ImageHashParseError` should implement `std::error::Error`.
```suggestion
            Self::UnparsableBytes(_) => f.write_str("Could not parse to hex"),
```

arqunis commented on src/model/misc.rs:
```diff
@@ -53,6 +51,108 @@ impl_from_str! {
+}
+
+impl std::fmt::Display for ImageHash {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```
Additionally, if `Display` will the the go-to way for 'peeking' into an `ImageHash`, it should be documented how the output should look like.

arqunis commented on src/model/misc.rs:
```diff
@@ -53,6 +51,108 @@ impl_from_str! {
+        } else if s.len() == 32 {
+            (s, false)
+        } else {
+            return Err(Self::Err::InvalidLength(s.len()));
```
Pedantic, but for consistency.
```suggestion
            return Err(ImageHashParseError::InvalidLength(s.len()));
```

GnomedDev committed:
Fix accidental signature change

Co-authored-by: Alex M. M. <acdenissk69@gmail.com>

GnomedDev committed:
Add handling for Clyde bot

This is untested as I do not have access to it.

GnomedDev committed:
Add proper documentation

GnomedDev committed:
Simplify error cases

This removes the weird case of an input of length 34 being a different error case

GnomedDev committed:
Document ImageHashParseError

GnomedDev committed:
Be consistent about Self::Err vs ImageHashParseError

GnomedDev committed:
Implement std::error::Error for ImageHashParseError

GnomedDev committed:
Fix no default features
